framework/db/ActiveRecord.php

```diff
@@ -199,34 +200,58 @@ protected static function findByCondition($condition)
      * This method will ensure that an array condition only filters on existing table columns.
      *
      * @param array $condition condition to filter.
+     * @param array $aliases
      * @return array filtered condition.
      * @throws InvalidArgumentException in case array contains unsafe values.
+     * @throws InvalidConfigException
      * @since 2.0.15
      * @internal
      */
-    protected static function filterCondition(array $condition)
+    protected static function filterCondition(array $condition, array $aliases = [])
     {
         $result = [];
         $db = static::getDb();
-        // valid column names are table column names or column names prefixed with table name
+        $columnNames = static::filterValidColumnNames($db, $aliases);
+
+        foreach ($condition as $key => $value) {
+            if (is_string($key) && !in_array($db->quoteSql($key), $columnNames, true)) {
+                throw new InvalidArgumentException('Key "' . $key . '" is not a column name and can not be used as a filter');
+            }
+            $result[$key] = is_array($value) ? array_values($value) : $value;
+        }
+
+        return $result;
+    }
+
+    /**
+     * Valid column names are table column names or column names prefixed with table name or table alias
+     *
+     * @param Connection $db
+     * @param array $aliases
+     * @return array
+     * @throws InvalidConfigException
+     */
+    protected static function filterValidColumnNames($db, array $aliases)
+    {
         $columnNames = [];
         $tableName = static::tableName();
         $quotedTableName = $db->quoteTableName($tableName);
+        $tableAliases = array_keys($aliases);
+        unset($tableAliases[$tableName]);
```

Comment:
This probably will not work - `$tableAliases` is now integer-indexed array, `$tableAliases[$tableName]` probably never exist.

Is there any reason we're not using regexp for this validation? Whitelist approach looks like a source of countless edge cases:
1. `$this->from` is not always associative, it could be `['table1', 'table2']` and then `array_keys($aliases)` will return array of integers. Does it make sense to support `0.id` as column name?
2. Right now every additional table name is also ignored, so for `['t1' => 'table1', 't2' => 'table2']` column `table2.id` will be reported as incorrect.

Reply:
I completely rewrote the alias handling and I wrote tests for it too.